### PR TITLE
[wip] enable settings kube-reserved cgroup

### PR DESCRIPTION
**Pull request by alexeldeib**
it's best practice to run kubelet + docker in a separate cgroup for kube-reserved processes. this enforces isolation from other system daemons under resource pressure. This PR makes doing so possible on top of https://github.com/Azure/aks-engine/pull/3530 which itself is a cherry-pick of https://github.com/Azure/aks-engine/pull/3201 to AKS.

### CODE REVIEW

File: parts/linux/cloud-init/artifacts/cse_config.sh [hunk 1]
```diff
@@ -268,6 +275,12 @@ ensureKubelet() {
     wait_for_file 1200 1 $KUBECONFIG_FILE || exit $ERR_FILE_WATCH_TIMEOUT
     KUBELET_RUNTIME_CONFIG_SCRIPT_FILE=/opt/azure/containers/kubelet.sh
     wait_for_file 1200 1 $KUBELET_RUNTIME_CONFIG_SCRIPT_FILE || exit $ERR_FILE_WATCH_TIMEOUT
+    {{- if HasKubeReservedCgroup}}
+    KUBERESERVED_SLICE_FILE=/etc/systemd/system/{{- GetKubeReservedCgroup -}}.slice
+    wait_for_file 1200 1 $KUBERESERVED_SLICE_FILE || exit $ERR_KUBERESERVED_SLICE_SETUP_FAIL
+    KUBELET_SLICE_FILE=/etc/systemd/system/kubelet.service.d/kubereserved-slice.conf
```
Comment (xuto2) [inline parts/linux/cloud-init/artifacts/cse_config.sh:281]: nit: s/KUBELET_SLICE_FILE/KUBELET_SLICE_CONFIG_FILE

File: parts/linux/cloud-init/artifacts/cse_config.sh [hunk 2]
```diff
@@ -226,6 +229,10 @@ ensureDocker() {
     if [[ $OS != $COREOS_OS_NAME ]]; then
         wait_for_file 1200 1 $DOCKER_MOUNT_FLAGS_SYSTEMD_FILE || exit $ERR_FILE_WATCH_TIMEOUT
     fi
+    {{- if HasKubeReservedCgroup}}
+    DOCKER_SLICE_FILE=/etc/systemd/system/docker.service.d/kubereserved-slice.conf
```
Comment (xuto2) [inline parts/linux/cloud-init/artifacts/cse_config.sh:233]: nit: s/DOCKER_SLICE_FILE/DOCKER_SLICE_CONF_FILE

File: pkg/agent/baker.go [hunk 3]
```diff
@@ -527,6 +527,18 @@ func getContainerServiceFuncMap(cs *api.ContainerService, profile *api.AgentPool
 		"GetDataDir": func() string {
 			return cs.Properties.OrchestratorProfile.KubernetesConfig.ContainerRuntimeConfig[common.ContainerDataDirKey]
 		},
+		"HasKubeReservedCgroup": func() bool {
+			kc := cs.Properties.OrchestratorProfile.KubernetesConfig
```
Comment (xuto2) [inline pkg/agent/baker.go:531]: Shall we allow this to be set per AgentPool?

File: parts/linux/cloud-init/artifacts/cse_config.sh [hunk 4]
```diff
@@ -207,14 +207,17 @@ ensureContainerRuntime() {
...
-  {{end}}
-  systemctl is-active --quiet docker && (systemctl_disable 20 30 120 docker || exit $ERR_SYSTEMD_DOCKER_STOP_FAIL)
-  systemctlEnableAndStart containerd || exit $ERR_SYSTEMCTL_START_FAIL
+    wait_for_file 1200 1 /etc/systemd/system/containerd.service.d/exec_start.conf || exit $ERR_FILE_WATCH_TIMEOUT
+    wait_for_file 1200 1 /etc/containerd/config.toml || exit $ERR_FILE_WATCH_TIMEOUT
+    {{if IsKubenet }}
+    wait_for_file 1200 1 /etc/sysctl.d/11-containerd.conf || exit $ERR_FILE_WATCH_TIMEOUT
+    retrycmd_if_failure 120 5 25 sysctl --system || exit $ERR_SYSCTL_RELOAD
+    {{end}}
+    {{- if HasKubeReservedCgroup}}
```
Comment (alexeldeib) [inline parts/linux/cloud-init/artifacts/cse_config.sh:216]: diff before this line is just +2 spaces for indentation consistency

File: cmd/generate.go [hunk 5]
```diff
@@ -234,7 +234,14 @@ func (gc *generateCmd) run() error {
 
 	customDataStr := templateGenerator.GetNodeBootstrappingPayload(gc.containerService, gc.containerService.Properties.AgentPoolProfiles[0])
 
-	cseCmdStr := templateGenerator.GetNodeBootstrappingCmd(gc.containerService, gc.containerService.Properties.AgentPoolProfiles[0], "<tenantid>", "<subid>", "rgname", "msiid")
+	cseCmdStr := templateGenerator.GetNodeBootstrappingCmd(gc.containerService, gc.containerService.Properties.AgentPoolProfiles[0],
```
Comment (eriksywu) [inline cmd/generate.go:237]: hmm how come I don't see this `GetNodeBootstrappingCmd` impl change in this PR?

Comment (alexeldeib) [reply cmd/generate.go:237]: the change isn't in my PR, the build was broken tho. I don't think this is used anywhere anyway

Comment (alexeldeib) [reply cmd/generate.go:237]: https://github.com/Azure/AgentBaker/commit/774d1d3b8755dd748faf016e2ec1a000721877ed#diff-a284f4316cf603f1f9b3cacaffc39a62R93

**Pull request closed without merging**